com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,344 @@
+            ChangedClassFilesVisitor visitor = getChangedClassFiles(event);
+            if (visitor != null) {
+                List<IResource> resources = visitor.getChangedClassFiles();
+                List<String> fullyQualifiedName = visitor.getQualifiedNamesList();
+                if (resources != null && !resources.isEmpty()) {
+                    doHotCodeReplace(resources, fullyQualifiedName);
+                } else {
+                    synchronized (needHotCodeReplace) {
+                        needHotCodeReplace.set(false);
+                        needHotCodeReplace.notifyAll();
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<List<String>> redefinedClasses() {
+        synchronized (needHotCodeReplace) {
+            try {
+                needHotCodeReplace.wait(100);
+                if (needHotCodeReplace.get()) {
+                    needHotCodeReplace.wait();
+                }
+            } catch (InterruptedException e) {
+                logger.log(Level.INFO, "The hotcode replace completion was interrupted by " + e.getMessage(), e);
+            }
+        }
+        return CompletableFuture.completedFuture(new ArrayList<String>());
+    }
+
+    private void doHotCodeReplace(List<IResource> resources, List<String> fullyQualifiedName) {
+        // DODO: Add HCR logic
+        synchronized (needHotCodeReplace) {
+            needHotCodeReplace.set(false);
+            needHotCodeReplace.notifyAll();
+        }
+    }
+
+    /**
+     * Returns the class file visitor after visiting the resource change. The
+     * visitor contains the changed class files and qualified type names. Returns
+     * <code>null</code> if the visitor encounters an exception, or the delta is not
+     * a POST_BUILD.
+     */
+    private ChangedClassFilesVisitor getChangedClassFiles(IResourceChangeEvent event) {
+        IResourceDelta delta = event.getDelta();
+        if (event.getType() != IResourceChangeEvent.POST_BUILD || delta == null) {
+            return null;
+        }
+        classFilesVisitor.reset();
+        try {
+            delta.accept(classFilesVisitor);
+        } catch (CoreException e) {
+            return null; // quiet failure
+        }
+        return classFilesVisitor;
+    }
+
+    /**
+     * Returns the class file or compilation unit containing the given fully
+     * qualified name in the specified project. All registered java like file
+     * extensions are considered.
+     *
+     * @param qualifiedTypeName
+     *            fully qualified type name
+     * @param project
+     *            project to search in
+     * @return class file or compilation unit or <code>null</code>
+     * @throws CoreException
+     *             if an exception occurs
+     */
+    public static IJavaElement findElement(String qualifiedTypeName, IJavaProject project) throws CoreException {
+        String path = qualifiedTypeName;
```

Comment:
Use below logic to implement finding element for fqn.

<pre>
        // lookup type
        NameLookup lookup = newNameLookup(DefaultWorkingCopyOwner.PRIMARY);
        NameLookup.Answer answer = lookup.findType(
                qualifiedTypeName,
            false,
            NameLookup.ACCEPT_ALL,
            true/* consider secondary types */,
            false/* do NOT wait for indexes */,
            false/*don't check restrictions*/,
            null);

        if (answer != null) {
            return answer.type.getParent();
        } else {
            return null;
        }
</pre>